File: tools/orchestrator/src/pipeline/buildPdf.ts
import fs from "node:fs/promises";
import path from "node:path";
import { spawn } from "node:child_process";

export type BuildPdfOpts = {
  texPath: string;
  pdfPath?: string;
  logPath?: string;
  repoRoot?: string;
};

function runCommand(command: string, args: string[], cwd: string): Promise<{ code: number; stdout: string; stderr: string }> {
  return new Promise((resolve, reject) => {
    const child = spawn(command, args, { cwd, shell: false });

    let stdout = "";
    let stderr = "";

    child.stdout.on("data", (chunk) => {
      stdout += chunk.toString();
    });
    child.stderr.on("data", (chunk) => {
      stderr += chunk.toString();
    });

    child.on("error", (err) => reject(err));
    child.on("close", (code) => resolve({ code: code ?? 1, stdout, stderr }));
  });
}

export async function buildPdf(opts: BuildPdfOpts): Promise<void> {
  const texAbs = path.resolve(opts.texPath);
  const pdfAbs = path.resolve(opts.pdfPath ?? texAbs.replace(/\.tex$/i, ".pdf"));
  const logAbs = opts.logPath ? path.resolve(opts.logPath) : undefined;

  const repoRoot = opts.repoRoot ? path.resolve(opts.repoRoot) : await findRepoRoot(process.cwd());
  const scriptPath = path.join(repoRoot, "scripts", "latex_to_pdf");

  const cwd = repoRoot;
  const args = [scriptPath, texAbs, pdfAbs];

  const { code, stdout, stderr } = await runCommand("bash", args, cwd);
  const log = [
    `command: bash ${args.map((a) => JSON.stringify(a)).join(" ")}`,
    `exit_code: ${code}`,
    "--- stdout ---",
    stdout.trimEnd(),
    "--- stderr ---",
    stderr.trimEnd(),
    ""
  ].join("\n");

  if (logAbs) {
    await fs.writeFile(logAbs, log, "utf8");
  }

  if (code !== 0) {
    throw new Error(`latex build failed (exit ${code}). See ${logAbs ?? "stderr"}.`);
  }
}

async function findRepoRoot(startDir: string): Promise<string> {
  let current = path.resolve(startDir);
  for (let i = 0; i < 10; i += 1) {
    const candidate = path.join(current, "scripts", "latex_to_pdf");
    try {
      await fs.access(candidate);
      return current;
    } catch {
      // keep walking
    }
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
  }
  return path.resolve(startDir);
}

File: backend/src/resume_orchestrator/pipeline/build_pdf.py
from __future__ import annotations

import asyncio
import json
from pathlib import Path


async def _run_command(command: str, args: list[str], cwd: Path) -> tuple[int, str, str]:
    proc = await asyncio.create_subprocess_exec(
        command,
        *args,
        cwd=str(cwd),
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    stdout_b, stderr_b = await proc.communicate()
    return proc.returncode or 0, stdout_b.decode("utf-8", errors="replace"), stderr_b.decode("utf-8", errors="replace")


def _find_repo_root(start_dir: Path) -> Path:
    current = start_dir.resolve()
    for _ in range(10):
        candidate = current / "scripts" / "latex_to_pdf"
        if candidate.exists():
            return current
        parent = current.parent
        if parent == current:
            break
        current = parent
    return start_dir.resolve()


async def build_pdf(
    *,
    tex_path: str | Path,
    pdf_path: str | Path | None = None,
    log_path: str | Path | None = None,
    repo_root: str | Path | None = None,
) -> None:
    tex_abs = Path(tex_path).resolve()
    pdf_abs = (
        Path(pdf_path).resolve()
        if pdf_path
        else tex_abs.with_suffix(".pdf")
    )
    log_abs = Path(log_path).resolve() if log_path else None
    root = Path(repo_root).resolve() if repo_root else _find_repo_root(Path.cwd())
    script_path = root / "scripts" / "latex_to_pdf"

    args = [str(script_path), str(tex_abs), str(pdf_abs)]
    code, stdout, stderr = await _run_command("bash", args, root)

    cmd_repr = "bash " + " ".join(json.dumps(a) for a in args)
    log = "\n".join(
        [
            f"command: {cmd_repr}",
            f"exit_code: {code}",
            "--- stdout ---",
            stdout.rstrip(),
            "--- stderr ---",
            stderr.rstrip(),
            "",
        ]
    )

    if log_abs is not None:
        log_abs.write_text(log, encoding="utf-8")

    if code != 0:
        location = str(log_abs) if log_abs else "stderr"
        raise RuntimeError(f"latex build failed (exit {code}). See {location}.")
